python-package/lightgbm/sklearn.py

```diff
@@ -635,6 +656,25 @@ def _process_params(self, stage: str) -> Dict[str, Any]:
         # overwrite default metric by explicitly set metric
         params = _choose_param_value("metric", params, original_metric)
 
+        # use joblib conventions for negative n_jobs, just like scikit-learn
+        n_jobs = self.n_jobs
+        if stage == "fit":
+            del params["n_jobs"]
+        num_threads_aliases = _ConfigAliases.get("num_threads")
+        for alias in num_threads_aliases:
+            if alias in params:
+                n_jobs = params.pop(alias)
+        if n_jobs is None or n_jobs < 0:
+            if n_jobs is None:
+                if cpu_count_lib == "joblib":
+                    params["num_threads"] = cpu_count(only_physical_cores=True)
+                elif cpu_count_lib == "psutil":
+                    params["num_threads"] = cpu_count(logical=False)
+                else:
+                    params["num_threads"] = cpu_count()
+            elif n_jobs < 0:
+                params["num_threads"] = max(cpu_count() + 1 + self.n_jobs, 1)
```

Comment:
I think this can be done with less lines of code:
```suggestion
        if stage == "fit":
            params.pop('n_jobs', None)
        params = _choose_param_value("num_threads", params, self.n_jobs)
        n_jobs = params["num_threads"]
        if n_jobs is None:
            params["num_threads"] = _LGBMCpuCount()
        elif n_jobs < 0:
            params["num_threads"] = max(_LGBMCpuCount() + 1 + n_jobs, 1)
```